Languages

Ana06 opened the pull request:
Issue #760

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,15 @@ def create_difficulty_levels
+    # All white spaces are removed to allow languages to be separated by ',' and ', ' but only the first option is saved
+    if self.languages.present?
+      self.languages = self.languages.delete(' ')
+      self.languages.match(/^$|(\A[a-z][a-z](,[a-z][a-z])*\z)/).present?
```
Hm, that means that I could also enter ``foobar`` as language?

Ana06 replied:
No, as the regular expression checks that it is a list of words of two letters. Although you could enter `fo` as language and it is not. I could split the string and check if every language is a real language.

ChrisBr replied:
Does it make sense to add more than one language? I mean usually the speaker presents only in one language (sure there could be subtitles, however, I'd neglect that ...)

Ana06 replied:
In the event only one language is selected, but this code is for the program, where the languages that are allowed for events are defined. Then it is a list of languages. There is a selector when you create a proposal with all the languages in this list

ChrisBr replied:
Ok

ChrisBr commented:
Your implemenation goes in the direction of labeling proposals. There is a nice gem for this: https://github.com/mbleigh/acts-as-taggable-on

Basically I like the idea, however, I don't like the self coded labeling and also there is no validation if the input is a language.

Also it should be possible to prohibit entering a language as Organizer!

ChrisBr commented:
So from my POV:
- A proposal / event has one language associated
- The Organizers should be able to prohibit selecting a different language, provide languages which the speakers can select (I don't want that the speaker can input every possible language, eg for a Conference in Germany the Organizers want to allow only German and English as presentation language) and also a default language!

ChrisBr commented:
Maybe this gem could help to solve the problem: https://github.com/scsmith/language_list

ChrisBr commented:
@hennevogel thoughts?

ChrisBr commented:
Also please provide screenshots

Ana06 commented:
A proposal/event has only one language of a list of languages defined by organizers. The languages are words of two letters as in ISO 639-1. It was suggested by @hennevogel in issue #760 to use ISO 639-1.

The program in the backend looks like that:

![program-edit](https://cloud.githubusercontent.com/assets/16052290/14003862/1b07cfe8-f156-11e5-98ba-ea27a37f5b27.jpg)

![program](https://cloud.githubusercontent.com/assets/16052290/14003880/3b1cd2f6-f156-11e5-8749-5446430ccae3.jpg)

And when you create a new proposal:

![proposal-new](https://cloud.githubusercontent.com/assets/16052290/14003901/76639282-f156-11e5-905f-a623b18d81ae.jpg)


There is not a default language. I think it should be a select of the languages in the list in the same place where the list is written. I can add it.

It is not checked if the languages introduced represent a real language. I also think that it could be useful to show `english` instead of `en` when creating a proposal (as most people won't know what `en` means). Both things can be done with language list gem, although it is said that it has functions to retrieve only common languages. Maybe it is better to have a hash with all codes and its language instead of using the gem. I think that using ActsAsTaggableOn gem is not very useful here, although using parse => true instead of split(',') is probably more elegant using add and remove won't be very nice, as every time languages field changes all tags should be removed and the new ones added (maybe not all of them are new, but it would be the same as checking if they are new or not it is not worthwhile). Maybe I'm wrong but I think that this is to simple to use ActsAsTaggableOn gem.

Ana06 commented:
Do the language should be displayed in any other place? I have added in:

![events-index](https://cloud.githubusercontent.com/assets/16052290/14004672/1b521c82-f15c-11e5-93a7-9f0dd5897809.jpg)

![events-show](https://cloud.githubusercontent.com/assets/16052290/14004673/1b5b96fe-f15c-11e5-930a-a0d700ebeb31.jpg)

ChrisBr commented:
@Ana06 sorry haven't had any time yet, will look into it on the weekend!

Ana06 commented:
Ok! :smile:

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,15 @@ def create_difficulty_levels
+  #
+  def check_languages_format
+    # All white spaces are removed to allow languages to be separated by ',' and ', ' but only the first option is saved
+    if self.languages.present?
```
Do you need the ``self``? If not, drop it!

Ana06 replied:
I think that `self` is needed

ChrisBr commented on app/controllers/proposal_controller.rb:
```diff
@@ -16,6 +16,8 @@ def show
   def new
     @user = User.new
     @url = conference_program_proposal_index_path(@conference.short_title)
+    @has_language = @program.languages.present?
```
Merge these two lines!

ChrisBr replied:
What about the :edit action?

Ana06 replied:
I defined @has_language because I also need to know if languages is present in the view, but I can change it.
I forgot the :edit action :frowning:

ChrisBr commented:
Also please fix the RuboCop offence
```
app/models/program.rb:104:5: C: Style/GuardClause: Use a guard clause instead of wrapping the code inside a conditional expression.
```

Ana06 commented:
I've made all corrections you have pointed out.
I've also introduced a text indicating that the fist language in the list is the default language (that was already implemented as the first language is the first shown in the selector).
I have found a gem (https://github.com/alphabetum/iso-639) which I think is worthwhile using. I'm using it to valid languages and to translate languages ('en' to 'English'). It looks like that:

![backend](https://cloud.githubusercontent.com/assets/16052290/14193243/7f12f992-f7a3-11e5-918c-db5248045efa.png)

![new_proposal](https://cloud.githubusercontent.com/assets/16052290/14193249/8b31b0b0-f7a3-11e5-8b37-9b278fcbfb8c.png)

ChrisBr commented on Gemfile.lock:
```diff
@@ -579,3 +581,6 @@ DEPENDENCIES
   web-console (~> 2.0)
   webmock
   whenever
+
```
Drop these lines please

ChrisBr commented on app/controllers/proposal_controller.rb:
```diff
@@ -16,11 +16,13 @@ def show
   def new
     @user = User.new
     @url = conference_program_proposal_index_path(@conference.short_title)
+    @languages = @program.languages.split(',').map {|l| ISO_639.find(l).english_name} if @program.languages.present?
```
Please do that in a function in the program model!

ChrisBr commented on app/controllers/proposal_controller.rb:
```diff
@@ -16,11 +16,13 @@ def show
   def edit
     authorize! :edit, @event
     @url = conference_program_proposal_path(@conference.short_title, params[:id])
+    @languages = @program.languages.split(',').map {|l| ISO_639.find(l).english_name} if @program.languages.present?
```
That's exactly the same line code as in the new action ... so please DRY it in a function!

Ana06 replied:
Sorry, I was going to move it to a function but I forgot. :open_mouth:

ChrisBr commented on app/models/program.rb:
```diff
@@ -41,6 +41,7 @@ def confirmed
 
   before_create :create_event_types
   before_create :create_difficulty_levels
+  before_save :check_languages_format
```
Why a before_save hook? What happens if it fails? Does the user get an error message? For me this looks more like a validation ...

Ana06 replied:
The user get a generic error message. I am going to use validation to include more specific error messages.

ChrisBr commented:
Please also rebase!

differentreality commented on Gemfile.lock:
```diff
@@ -580,9 +582,7 @@ DEPENDENCIES
   unobtrusive_flash (>= 3)
-  web-console (~> 2.0)
-  webmock
-  whenever
```
Why are we removing these 3 lines? Did you do that manually?
I think @ChrisBr 's comment might have been about the lines below this (lines 585-588), as the comment was on the empty line, not on the ones you removed. 
In any case, you cannot remove gems from the Gemfile.lock while you still have them in your Gemfile. That's what's causing travis to fail.

Ana06 replied:
I did it manually. I though that it was strange to remove them and I saw that travis failed because of that. I clearly misunderstand what @ChrisBr said. I'll fix it. Thank you. :wink:

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,29 @@ def create_difficulty_levels
   end
+
+  ##
+  # Check if languages string has the right format. Used as before_save.
```
The comment is wrong!

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,29 @@ def create_difficulty_levels
+    errors.add(:languages, 'must be two letters separated by commas') && return unless
+    self.languages.match(/^$|(\A[a-z][a-z](,[a-z][a-z])*\z)/).present?
+    error = false
+    self.languages.split(',').each do |lan|
```
I just looked briefly into it but you could rewrite it as

```
self.languages.split(',').each do |lan|
  if ISO_639.find(lan).nil?
     errors.add(:languages, 'must be ISO 639-1 valid codes')
     return
end
```
Then you don't need the variable and the extra check after the loop!

vicgalle replied:
I had done it like that but Rubocop doesn't like it:
http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/NonLocalExitFromIterator

Ana06 replied:
I had done it like that but Rubocop doesn't like it:
http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/NonLocalExitFromIterator

ChrisBr replied:
Then break out of the loop at least!

differentreality replied:
Why don't we substitute all that loop just with
`errors.add(:languages, 'must be ISO 639-1 valid codes') if self.languages.split(',').select{ |x| ISO_639.find(x).nil? }`

Ana06 replied:
@differentreality I think that you solution is much more elegant, although `select{}` will return an array (empty or not) and it will always add the error. But we could check if the array is not empty:

`errors.add(:languages, 'must be ISO 639-1 valid codes') unless self.languages.split(',').select{ |x| ISO_639.find(x).nil? }.empty?`

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,29 @@ def create_difficulty_levels
+    error = false
+    self.languages.split(',').each do |lan|
+      error = ISO_639.find(lan).nil?
+      next if !error
```
Ah and furthermore, this won't work! It iterates anyway over the whole Enumerable, what you would need is 
``
break if error
``
(but anyway, I think my other solution would be better)

ChrisBr replied:
As mentioned, I just looked briefly into it, maybe there is another better solution! Will have a deeper look tomorrow!

Ana06 replied:
I also think it is better but I am not sure if Rubocop would like it.

ChrisBr replied:
Why not? Break should be totally valid!

Ana06 replied:
I meant that Rubocop wouldn't like your other solution. I think Rubocop would like `break`.

ChrisBr commented on app/models/program.rb:
```diff
@@ -94,4 +95,29 @@ def create_difficulty_levels
+    end
+  end
+
+  def languages_list
```
just ```languages``` should be enough!

vicgalle replied:
I am not pretty sure what you mean

ChrisBr commented on app/views/admin/events/index.html.haml:
```diff
@@ -38,6 +41,7 @@
         %tr
           %td
             = event.id
+            
```
What about this whitespace change?

Ana06 replied:
I though It was more readable with it. I can delete it if you want.

ChrisBr commented on app/views/admin/events/index.html.haml:
```diff
@@ -70,6 +74,7 @@
             - bgcolor="#F7819F"
           - else
             - bgcolor=""
+
```
What about this whitespace change?

ChrisBr commented on app/views/admin/programs/_form.html.haml:
```diff
@@ -8,5 +8,6 @@
       = f.input :schedule_public, label: "Show Schedule on the home and splash page"
       = f.input :schedule_fluid, label: "Allow submitters to change their event after it is scheduled"
       = f.input :rating, :hint => "Enter the number of different rating levels you want to have for voting on proposals. Enter 0 if you do not want to vote on proposals."
+      = f.input :languages, :hint => "Enter the languages allowed for events as values of ISO 639-1 language codes separated with commas. The first language would be the default language. Leave it blank if you do not want to specify languages."
```
What about a link to a description of the ISO-639?

Ana06 replied:
I'll add it. I suppose you mean this link:

http://www.loc.gov/standards/iso639-2/php/code_list.php

ChrisBr commented:
@hennevogel please also have a look, it was your idea ;)

hennevogel commented:
@ChrisBr I'm fine with your review :-)

ChrisBr commented:
@Ana06 any update?

Ana06 commented:
Sorry, I couldn't solve it before. I was really busy this week.

I have eliminated the white spaces I added (I had added because I though the code was more readable).

I have changed `next if !error` for `break if error`. I thought both things do the same. In fact, I had never used break or next as I had always used the other way Rubocop doesn't like with return. It seems that Rubocop is happy with that.

I solved a problem with languages_list function.

Last, I added the link to ISO-639.

I hope it is okey now. :smile:

Ana06 commented:
@ChrisBr @hennevogel is there any other change I should make?

hennevogel commented:
@Ana06 please rebase

Ana06 commented:
Done! :smile:  Why does Travis CI fail?

differentreality commented:
@Ana06 it previously failed due to phantomjs, I restarted travis and now it has a real error. Please take a look?

differentreality commented:
@Ana06 The second style error must be the same issue travis has in general. 
Please also rebase again against the latest changes from yesterday.

On a different note, I am not a big fan of organizers entering strings for language manually. Couldn't we rework that into a multiple choice option? With prefilled text.

From the proposal submitter's point of view I think that something like **EN - English**, instead of just EN, or even just **English**, would make it even more straight forward.
If we stick with just initials, it would also be nice to show EN instead of en, I think that's the common way to write language initials, and some styling is nice.
Moreover if we ever wanted to know any laguage statistics accross the conferences we had, it will be impossible when we use different indicators.

ChrisBr commented:
@Ana06 also please rebase!

Ana06 commented:
@differentreality I don't understand how to use multiple choice option as there are a lot of possible languages.


I have reviewed the code and I realised that capital letters where not allowed when introducing languages in the backend. I have changed it to allow then, although I save then in the same way. I mean, the following options are all allowed:

`en,de,es`
`en, de, es`
`EN,DE,ES`
`EN, de, ES`

But in the four cases `en,de,es` is saved. So there is not any problem with statistics, as English is always saved as `en`.

![backend](https://cloud.githubusercontent.com/assets/16052290/14967142/3697e3d6-10b6-11e6-9172-bd3703181bf6.png)



And that is what proposal submitters see:
![submitters](https://cloud.githubusercontent.com/assets/16052290/14967651/bd17e5bc-10b8-11e6-8713-de68fd977766.png)
I think that it already does what you have suggested (it shows `English`), or maybe I understood you wrong.

differentreality commented on app/views/proposal/_proposal_form.html.haml:
```diff
@@ -13,7 +13,12 @@
       include_blank: false, label: 'Type', input_html: { class: 'select-help-toggle' }
+      - if @program.languages.present?
 
+    = f.input :language, as: :select,
```
Identation doesn't seem right

differentreality commented on app/views/admin/programs/_form.html.haml:
```diff
@@ -8,5 +8,6 @@
       = f.input :schedule_fluid, label: "Allow submitters to change their event after it is scheduled"
       = f.input :rating, :hint => "Enter the number of different rating levels you want to have for voting on proposals. Enter 0 if you do not want to vote on proposals."
+      = f.input :languages, :hint => "Enter the languages allowed for events as values of #{link_to('ISO 639-1', 'http://www.loc.gov/standards/iso639-2/php/code_list.php', :target => "_blank")} language codes separated with commas. The first language would be the default language. Leave it blank if you do not want to specify languages.".html_safe
       %p.text-right
```
Please use the new ruby style

Ana06 replied:
What do you mean with _new ruby style_?

differentreality replied:
@Ana06 There is a new coding style for ruby, which we try to follow in new code we write. Googling about it: https://github.com/bbatsov/ruby-style-guide#hash-literals

Ana06 replied:
Ok! I'll change it

Ana06 commented:
I have changed to the new ruby style the whole `app/views/admin/programs/_form.html.haml` file

differentreality commented:
@Ana06 for future reference, it is enough to make sure you use the new ruby style for the new code you introduce. Otherwise it may get difficult to review both styling changes and new feature implementation. But you can always open a new PR just with styling changes ;)

You should also use single quotes, when you don't need string interpolation.

differentreality commented on app/models/program.rb:
```diff
@@ -113,4 +118,23 @@ def create_difficulty_levels
+  # Check if languages string has the right format. Used as validation.
+  #
+  def check_languages_format
+    # All white spaces are removed to allow languages to be separated by ',' and ', ' but only the first option is saved
```
This comment should be one line lower, as it refers to L128, not L127
What do you mean only the first option is saved?

Ana06 replied:
I mean that `en, de` and `en,de` are both allowed, but it is always saved as `en,de`, removing spaces (I did it like that for being consistent). I will change the comment to make it more clear.

differentreality commented on app/models/program.rb:
```diff
@@ -113,4 +118,23 @@ def create_difficulty_levels
+    # All white spaces are removed to allow languages to be separated by ',' and ', ' but only the first option is saved
+    return unless self.languages.present?
+    self.languages = self.languages.delete(' ').downcase
+    # We check if every language is a valid ISO 639-1 language
```
This line does not say what happens to the following line! That comment should be above the line it describes.

Ana06 replied:
I'll revise all comments in this function.

Ana06 commented:
I have just realised that languages could be repeated. I mean `en,de,en` was accepted. I suppose that we want to avoid that as it would be more difficult to know any language statistics and it would be confusing for the users that register for the conference too.

I have added this check and fixed all things @differentreality told me.

I won't make styling changes while I am implementing new features and I will use single quotes if possible from now on. :wink:

ChrisBr commented:
@Ana06 It looks good to me :+1: 

However, I'd wait till we released version 1.0 of OSEM before merging this as we don't want to introduce new features before 1.0! Thanks

Ana06 commented:
Ok! :wink:

Should I add any test?

hennevogel commented:
Always add tests :-)

Ana06 commented:
Ok :wink:

I've started adding tests, but I will add more tests. :smile:

Ana06 commented:
I've added tests in all test files that exists. I don't know if there is any reason to have no tests for some parts. Even If that parts need to have tests I think that it is better not to add them in this pull request to avoid confusion.

ChrisBr commented on spec/models/program_spec.rb:
```diff
@@ -58,4 +58,31 @@
   end
+
+  describe 'validation' do
+    it '#check_languages_format' do
```
@Ana06 please write for each test case a separate test because otherwise it's hard to debug if one test fails!

ChrisBr commented:
@Ana06 what is the default language if the user doesn't select one?

ChrisBr commented:
And please squash your commits!

Ana06 commented:
It is not possible to not select one. The selector has the first language by default, and there is no option to select no language. It could happen that the languages are set after the event is created, in that case the language would be `nil`.
By squash do you mean that I add more things in the same commit instead of doing several commits?

Ana06 commented:
I've squashed the commits and also split the long test. :blush:

ChrisBr commented on db/schema.rb:
```diff
@@ -351,9 +352,13 @@
     t.boolean  "include_registrations"
     t.boolean  "include_sponsors"
     t.boolean  "include_lodgings"
+    t.string   "banner_photo_file_name"
```
Huh?

Please drop these changes!

ChrisBr commented:
@hennevogel please review! From my side it's good to go!

hennevogel commented:
If you're okay it's okay
